Develop

mrucker opened the pull request:
You'll notice that there are now 6 project endpoints.

1. Android
2. Android Tests
3. iOS
4. iOS Tests
5. Local Tests
6. Cloud Tests

"Cloud Tests" is the old UI Tests project. I'm imagining we'll roll Android Tests and iOS Tests into Cloud Tests once we have time to get XTC running. That would l eave us with only four end points to manage then.

MatthewGerber commented:
Looking good. A couple things:

* Can we move tests from `Sensus.Tests.Local` to `Sensus.Shared.Tests`?
* I am concerned about maintaining identical dependencies in the `Sensus.Android` and `Sensus.Android.Tests` projects. Right now they each reference packages independently, so how do we guarantee that the deployment environment for the testing app is the same as for the production app? The testing app cannot reference the deployment app to get its dependencies, so what if we did something like this:  Add a class library project `Sensus.Android.App` that references everything that `Sensus.Android` currently does. When built, `Sensus.Android.App` will put all required dependencies in its output path. Then `Sensus.Android` and `Sensus.Android.Tests` would each reference `Sensus.Android.App` to get all of the dependencies needed to deploy. Each would be drawing from the same place (`Sensus.Android.App`) so the resulting deployment environments would be identical. The testing app would add the NUnit dependencies, which would be the only point of divergence between the two. Would something like this work?

mrucker commented:
- Yeah totally, I know we'd discussed that I just wanted to get an intermediate state out there. 
- As for the your reference idea I think that would work. I'll do it tomorrow and see how it goes.

mrucker commented:
- Also,

mrucker commented:
whoops wrong button...

mrucker commented:
Alright, I just moved all the tests into the shared test folder. I confirmed that the Sensus.Tests.Local still work. However, once I started running the tests on Android everything started crashing. Still looking into it.

You can go ahead and look at the project structure though. I don't think I like it as much. I moved the shared projects up into their respective platforms and then remained Sensus.Shared to Sensus.

As for the shared reference project I was thinking something like this.

- Sensus.Android
- Sensus.Android.App
- Sensus.Android.Tests

mrucker commented:
Alright,

Fixed the Android Unit Tests. Here's a fun tidbit for you both. On Android, if an undisposed class throws an exception during the finalizer the entire app dies. This is very different from .NET on Windows where exceptions during finalizers are very forgiving.

Looks like you already suggested the .App name. Thinking about it a little more, I think my final vote would be to keep it as is. Keeping those two projects in sync is simply something we're going to have to do. Otherwise I feel like we are kind of going back to the old PCL structure we had. At the very least I might suggest we let our changes sit for a few days or weeks. After we've had a chance write code against our new project structure we may have different opinions on where the pain points are.

MatthewGerber commented:
Is this ready for me to review?

mrucker commented:
Yeah the last few commits were simply me moving stuff around to see how  it looked.

mrucker commented on Sensus.Shared.Tests/Shared/ProtocolTests.cs:
```diff
@@ -0,0 +1,98 @@
+        #endregion
+
+        #region SetUp
+        [TestFixtureSetUp]
```
Deprecated? Also, should I make those namespace changes now that we have a project structure?

mrucker replied:
I see now. Fixed.

MatthewGerber commented on Sensus.Android/AndroidSensusServiceHelper.cs:
```diff
@@ -833,5 +829,15 @@ public override void BringToForeground()
+
+        public SensorManager GetSensorManager()
+        {
+            throw new NotImplementedException();
```
Bug, right? Should be implemented as below-left in AnroidSensorListener. I'll fix it.

mrucker replied:
Oh yeah. Whoops. Nice catch. Just as an FYI, this is mostly a stop gap change. Once we've managed to work out a lot of the dependencies and move AndroidSensusServiceHelper into the shared android project this won't be necessary.

mrucker commented:
Ha ok,

So the story with NUnit is that I've got an NUnit code runner that gives me a lot of nice functionality like code coverage and other bells and whistles. Unfortunately I've only ever been able to get it to work with NUnit 2.X. Any Library that is built against NUnit 3.X causes strange errors. The developer that maintains NUnit likes to tinker with the code a lot. My suspicion is that the developers of my test runner are tired of changing their code every time he releases a breaking change on a whim.

mrucker commented:
Oh so that's the reason I haven't brought Test.Local up to the latest version of NUnit. The reason I haven't brought the NUnit platform specific apps down to version 2.6.4 is because the comments with those projects say they rely in some way on NUnit 3.X functionality. I just took them at their word and never tried to bring it down.

mrucker commented on Sensus.Shared/DataStores/Local/LocalDataStore.cs:
```diff
@@ -214,7 +214,9 @@ public int WriteDataToZipFile(string zipPath, CancellationToken cancellationToke
-                    if (zipFile != null)
+#if __ANDROID__ || __IOS__
+                    if (zipFile != null)
+#endif
```
zipFile is only created above if we're in an Android or IOS build. As such this code wouldn't build in Sensus.Tests.Local without the if around zipFile. I changed it a little so that it is maybe a little more clear what is going on.

mrucker commented:
I just pushed a few more changes that better handle those iOS tests. I changed the failing iOS code so that it runs on the MainThreadSynchronizer and make sure to set that to the iOS specific synchronizer for our test environment.
